pkg/gitrepository/manager.go

```diff
@@ -0,0 +1,139 @@
+package gitrepository
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	profilesv1 "github.com/weaveworks/profiles/api/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//Manager is responsible for managing gitrepository resources
+type Manager struct {
+	kClient   Kubernetes
+	namespace string
+	ctx       context.Context
+	timeout   time.Duration
+	interval  time.Duration
+}
+
+//Instance contains a tag and path of profile.yaml
+type Instance struct {
+	Tag  string
+	Path string
+}
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate -o fakes/fake_kubernetes.go . Kubernetes
+//Kubernetes interface for itneracting with kubernetes
+type Kubernetes interface {
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error
+	Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error
+}
+
+//NewManager returns a Manager struct
+func NewManager(ctx context.Context, namespace string, kClient Kubernetes, timeout, interval time.Duration) *Manager {
+	return &Manager{
+		kClient:   kClient,
+		namespace: namespace,
+		ctx:       context.TODO(),
+		timeout:   timeout,
+		interval:  interval,
+	}
+}
+
+//CreateAndWaitForResources creates the gitrepository resources and waits for them to be created
+func (m *Manager) CreateAndWaitForResources(r profilesv1.Repository, instances []Instance) ([]*sourcev1.GitRepository, error) {
+	var gitResources []*sourcev1.GitRepository
+	for _, instance := range instances {
+		gitRes := makeGitRepository(r, instance.Tag, instance.Path, m.namespace)
+		err := m.kClient.Create(m.ctx, gitRes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create gitrepository: %w", err)
+		}
+		gitResources = append(gitResources, gitRes)
+	}
+
+	for _, gitRes := range gitResources {
+		err := m.waitForURL(gitRes)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return gitResources, nil
+}
+
+func (m *Manager) waitForURL(gitRes *sourcev1.GitRepository) error {
+	startTime := time.Now()
+	for {
+		err := m.kClient.Get(m.ctx, client.ObjectKeyFromObject(gitRes), gitRes)
+		if err != nil {
+			return fmt.Errorf("failed to get gitrepository: %w", err)
+		}
+		if gitRes.Status.URL != "" {
+			return nil
+		}
+
+		if time.Since(startTime) > m.timeout {
+			return fmt.Errorf("timed out waiting for %s/%s gitrepository.Status.URL to be populated", gitRes.Namespace, gitRes.Name)
+		}
+		time.Sleep(m.interval)
```

Comment:
I would much prefer using a time.After which doesn't waste as much resources as a regular sleep. That would look like this:

```go
ticker := time.NewTicker(10 * time.Second)
defer ticker.Stop()
for {
		select {
                 case <-ticker.C:
                      // do the url things
		case <-time.After(timeout):
                      // return error timeout.
                 }
}
```

Also gives much better control over timing out. And also doesn't mock around with time out after and things like that. And uses channel mechanics which are easier on the cpu than plain sleep.

Reply:
doesn't this mean we would hammer the k8s API? I introduced the delay on purpose, to prevent throttling.

Happy to change it to use a select statement all the same, but having a sleep after the k8s api call seems unavoidable?

Reply:
No, it doesn't. The select blocks. It's not going to call the k8s api continuously. But the ticker sucks I decided. The context with timeout is a LOT nicer. So use this instead which is much more readable if you like:

```go
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	for {
		select {
		case <-ctx.Done():
			fmt.Println("timeout")
			return err
		case <-time.After(interval):
			// do the thing in this interval
		}
	}
```

Example code: https://play.golang.org/p/SH8_uW9cUHy

Reply:
like it!